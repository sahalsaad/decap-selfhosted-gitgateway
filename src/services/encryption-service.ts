import nacl from "tweetnacl";
import util from "tweetnacl-util";
import crypto from "node:crypto";

const separator = "::";

const encrypt = (secret: string, encryptKey: string): string => {
  const keyBytes = util.decodeUTF8(encryptKey);
  const nonce = nacl.randomBytes(24);
  const encryptedSecret = nacl.secretbox(
    util.decodeUTF8(secret),
    nonce,
    keyBytes,
  );

  return `${util.encodeBase64(encryptedSecret)}${separator}${util.encodeBase64(nonce)}`;
};

const decrypt = (encryptedSecret: string, encryptKey: string): string => {
  const [secretString, nonce] = encryptedSecret.split(separator);
  const encryptedSecretBytes = util.decodeBase64(secretString);
  const nonceBytes = util.decodeBase64(nonce);
  const keyBytes = util.decodeUTF8(encryptKey);
  const secret = nacl.secretbox.open(
    encryptedSecretBytes,
    nonceBytes,
    keyBytes,
  );
  if (!secret) {
    throw new Error("Invalid encrypted secret");
  }
  return util.encodeUTF8(secret);
};

const hashPassword = (password: string, secretKey: string): string => {
  const secretKeyUint8Array = util.decodeUTF8(secretKey);
  const hmac = crypto.createHmac("sha256", secretKeyUint8Array);
  hmac.update(password);
  return hmac.digest("hex");
};

export { encrypt, decrypt, hashPassword };
