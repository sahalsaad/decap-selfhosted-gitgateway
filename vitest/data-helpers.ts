import { faker } from '@faker-js/faker'
import { hashPassword } from '@services/encryption-service'
import { env } from 'cloudflare:test'
import { sign } from 'hono/jwt'
import type { UserCreateRequest } from '@/types/user'

const generateSiteRequest = () => ({
  cmsUrl: faker.internet.url(),
  gitToken: faker.string.uuid(),
  gitRepo: faker.internet.url(),
  gitProvider: faker.helpers.arrayElement(['github', 'gitlab', 'bitbucket']),
  gitHost: faker.internet.url(),
})

const generateUserCreateRequest = (): UserCreateRequest => ({
  email: faker.internet.email(),
  firstName: faker.person.firstName(),
  lastName: faker.person.lastName(),
  password: faker.internet.password(),
  role: faker.helpers.arrayElement(['admin', 'contributor']),
})

const MOCK_ENV = {
  DB: env.DB,
  AUTH_SECRET_KEY: faker.string.uuid(),
}

const fakeAdminToken = await sign(
  { user: { id: 'fake-user-id', role: 'admin' } },
  MOCK_ENV.AUTH_SECRET_KEY
)

const fakeContributorToken = await sign(
  { user: { id: 'fake-user-id', role: 'contributor' } },
  MOCK_ENV.AUTH_SECRET_KEY
)

const fakeUserAndSiteData = (password: string) => ({
  id: faker.string.uuid(),
  email: faker.internet.email(),
  firstName: faker.person.firstName(),
  lastName: faker.person.lastName(),
  role: 'contributor',
  password: hashPassword(password, MOCK_ENV.AUTH_SECRET_KEY!),
  sites: [
    {
      id: faker.string.uuid(),
      gitToken: faker.string.uuid(),
      gitRepo: faker.internet.url(),
      gitProvider: faker.helpers.arrayElement(['github', 'gitlab', 'bitbucket']),
      gitHost: faker.internet.url(),
    },
  ],
})

export {
  generateSiteRequest,
  generateUserCreateRequest,
  MOCK_ENV,
  fakeAdminToken,
  fakeContributorToken,
  fakeUserAndSiteData,
}
