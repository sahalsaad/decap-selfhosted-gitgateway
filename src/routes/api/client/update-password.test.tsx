import { faker } from '@faker-js/faker'

import updatePasswordRoute from '@/src/routes/api/client/update-password'
import { MOCK_ENV, mockAdminToken } from '@/vitest/data-helpers'
import { hashPassword } from '@services/encryption-service'

const mockUserService = {
  getUserByEmail: vi.fn(),
  setPassword: vi.fn(),
}

vi.mock('@services/user-service', () => ({
  UserService: vi.fn().mockImplementation(() => mockUserService),
}))

describe('update password', () => {
  it('should return user not found if wrong email', async () => {
    mockUserService.getUserByEmail.mockResolvedValue(undefined)

    const response = await updatePasswordRoute.request(
      '/',
      {
        method: 'POST',
        headers: {
          'Authorization': `Bearer ${mockAdminToken}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          email: faker.internet.email(),
          currentPassword: faker.internet.password(),
          newPassword: faker.internet.password(),
        }),
      },
      MOCK_ENV,
    )
    expect(await response.text()).toMatchSnapshot()
  })

  it('should return incorrect password if invalid existing password', async () => {
    mockUserService.getUserByEmail.mockResolvedValue({
      id: faker.string.uuid(),
      email: faker.internet.email(),
      firstName: faker.person.firstName(),
      lastName: faker.person.lastName(),
      role: faker.helpers.arrayElement(['admin', 'contributor']),
      password: hashPassword(faker.internet.password(), MOCK_ENV.AUTH_SECRET_KEY!),
    })

    const response = await updatePasswordRoute.request(
      '/',
      {
        method: 'POST',
        headers: {
          'Authorization': `Bearer ${mockAdminToken}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          email: faker.internet.email(),
          currentPassword: faker.internet.password(),
          newPassword: faker.internet.password(),
        }),
      },
      MOCK_ENV,
    )
    expect(await response.text()).toMatchSnapshot()
  })

  it('should return errors if wrong request payload', async () => {
    const response = await updatePasswordRoute.request(
      '/',
      {
        method: 'POST',
        headers: {
          'Authorization': `Bearer ${mockAdminToken}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({}),
      },
      MOCK_ENV,
    )
    expect(await response.text()).toMatchSnapshot()
  })

  it('should return success if password updated', async () => {
    const password = faker.internet.password()
    mockUserService.getUserByEmail.mockResolvedValue({
      id: faker.string.uuid(),
      email: faker.internet.email(),
      firstName: faker.person.firstName(),
      lastName: faker.person.lastName(),
      role: faker.helpers.arrayElement(['admin', 'contributor']),
      password: hashPassword(password, MOCK_ENV.AUTH_SECRET_KEY!),
    })
    mockUserService.setPassword.mockResolvedValue(true)

    const response = await updatePasswordRoute.request(
      '/',
      {
        method: 'POST',
        headers: {
          'Authorization': `Bearer ${mockAdminToken}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          email: faker.internet.email(),
          currentPassword: password,
          newPassword: faker.internet.password(),
        }),
      },
      MOCK_ENV,
    )
    expect(await response.text()).toMatchSnapshot()
  })
})
