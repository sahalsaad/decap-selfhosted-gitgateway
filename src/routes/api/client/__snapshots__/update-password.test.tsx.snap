// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`update password > should return errors if wrong request payload 1`] = `"<span class="text-red-700">email: Required, currentPassword: Required, newPassword: Required</span>"`;

exports[`update password > should return incorrect password if invalid existing password 1`] = `"<span class="text-red-700">Incorrect password</span>"`;

exports[`update password > should return success if password updated 1`] = `"<div class="text-center text-2xl text-green-700">Password updated successfully!</div>"`;

exports[`update password > should return user not found if wrong email 1`] = `"<span class="text-red-700">User not found</span>"`;
